smtp/smtp.js

```diff
@@ -382,58 +476,117 @@ class SMTP extends EventEmitter {
+	/**
+	 * SMTP 'help' command, returns text from the server
+	 * @param {function(...*): void} callback function to call after response
+	 * @param {string} domain the domain to associate with the 'help' request
+	 * @returns {void}
+	 */
+	help(callback, domain) {
+		this.command(domain ? `help ${domain}` : 'help', callback, [211, 214]);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	rset(callback) {
 		this.command('rset', callback);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	noop(callback) {
 		this.send('noop', callback);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @param {string} from the sender
+	 * @returns {void}
+	 */
 	mail(callback, from) {
 		this.command(`mail FROM:${from}`, callback);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @param {string} to the receiver
+	 * @returns {void}
+	 */
 	rcpt(callback, to) {
 		this.command(`RCPT TO:${to}`, callback, [250, 251]);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	data(callback) {
 		this.command('data', callback, [354]);
 	}
 
+	/**
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	data_end(callback) {
 		this.command(`${CRLF}.`, callback);
 	}
 
+	/**
+	 * @param {string} data the message to send
+	 * @returns {void}
+	 */
 	message(data) {
 		log(data);
 		this.sock.write(data);
 	}
 
+	/**
+	 * SMTP 'verify' command -- checks for address validity.
+	 *
+	 * @param {string} address the address to validate
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	verify(address, callback) {
-		// SMTP 'verify' command -- checks for address validity.
 		this.command(`vrfy ${address}`, callback, [250, 251, 252]);
 	}
 
+	/**
+	 * SMTP 'expn' command -- expands a mailing list.
+	 *
+	 * @param {string} address the mailing list to expand
+	 * @param {function(...*): void} callback function to call after response
+	 * @returns {void}
+	 */
 	expn(address, callback) {
-		// SMTP 'expn' command -- expands a mailing list.
 		this.command(`expn ${address}`, callback);
 	}
 
+	/**
+	 * Calls this.ehlo() and, if an error occurs, this.helo().
+	 *
+	 * If there has been no previous EHLO or HELO command self session, self
+	 * method tries ESMTP EHLO first.
+	 *
+	 * @param {function(...*): void} callback function to call after response
+	 * @param {string} [domain] the domain to associate with the command
+	 * @returns {void}
+	 */
 	ehlo_or_helo_if_needed(callback, domain) {
-		// Call this.ehlo() and/or this.helo() if needed.
-		// If there has been no previous EHLO or HELO command self session, self
-		//  method tries ESMTP EHLO first.
+		// is this code callable...?
```

Comment:
i think this can be called from `this.login` if `this.ehlo` hasn't been run, since `this.ehlo` is what initializes an empty object for `this.features`. not entirely sure about the control flow, but it does actually appear to be callable.